Feature: support barcode upscaling for better detection of small barcodes

bmachek opened the pull request:
## Proposed change

My QR code labels weren't recognized during ASN barcode detection in paperless, as they seemed to be to small for pyzbar to detect them. I cannot use zxing, as I run paperless on arm64. Couldn't even test zxing, because I currently do not own a x86 device ;-)

I figured that converting the PDF pages with a dpi of 600 and then upscale the resulting image by 1.5 fixes my issue.


## Type of change

I added the ability to paperless to configure an upscale factor and the dpi used in the conversion.
Default values go to 1.0 upscale factor and dpi of 300, so nothing changes for users that are unaffected by this issue. The code won't try to upscale the image by 1.0, it skips the step.


- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

bmachek committed:
This adds the ability to configure if a PDF document is being upscaled and at what dpi it is converted to an image while detecting ASN barcodes.
This improves the detection of small qr codes with pyzbar in certain cases.

bmachek committed:
Added the two new config variables PAPERLESS_CONSUMER_ASN_BARCODE_UPSCALE and PAPERLESS_CONSUMER_ASN_BARCODE_DPI to documentation

Ran black over code changes in barcodes.py and settings.py

shamoon commented:
Thanks for the PR.

Yea there have been issues with this in the past, eg [#2516](https://github.com/paperless-ngx/paperless-ngx/issues/2516). I don't use the barcode feature, nor have I spent much time with that part of our code. 

In general this looks OK to me but Im always a little hesitant to add more config options, we're sort of at config overload already. Since we dont track anything, we dont have a good way to track how many people want / would use / would benefit from something like this (also a thought to opening discussions before PRs). But I do appreciate the 'sane' defaults and that this would leave things as-is for existing users.

Perhaps other folks will have more thoughts.

One other thing is that we like to include testing of new code, so if you could add that, eg https://github.com/paperless-ngx/paperless-ngx/blob/dev/src/documents/tests/test_barcodes.py

stumpylog commented:
I'm alright with adding the settings.  Though I do know we're very, very configurable at this point.

bmachek commented:
Ok, thanks for the feedback.
@shamoon Will add some tests to test_barcodes.py soon.

bmachek committed:
Wrote a pytest for barcode detection with upscaling and custom dpi value.
Added src/documents/tests/samples/barcodes/barcode-qr-asn-000123-upscale-dpi.pdf as sample for the test

Renamed CONSUMER_ASN_BARCODE_UPSCALE to CONSUMER_BARCODE_UPSCALE
Renamed CONSUMER_ASN_BARCODE_DPI to CONSUMER_BARCODE_DPI
in settings.py, barcodes.py, configuration.md and paperless.conf.example
Because the values apply in both ASN and seperate pages detection, these names are more appropriate

Changed CONSUMER_BARCODE_UPSCALE default value 0.0
Changed behavior in barcodes.py concerning upscaling. It will only take place with values bigger than 1.0
Downscaling and scaling by 1.0 don't make sense in this context. A debug message will be logged if upscaling happens.

bmachek commented on src/documents/barcodes.py:
```diff
@@ -203,11 +203,18 @@ def detect(self) -> None:
 
             for current_page_number, page in enumerate(pages_from_path):
+                factor = settings.CONSUMER_ASN_BARCODE_UPSCALE
+                if factor != 1.0:
```
Scaling by 1.0 didn't was already ruled out by the old code. In the new commit downscaling is also disabled, as I can't think of a use case.

bmachek commented on docs/configuration.md:
```diff
@@ -1095,6 +1095,22 @@ barcode.
 
     Defaults to "ASN"
 
+`PAPERLESS_CONSUMER_ASN_BARCODE_UPSCALE=<float>`
```
Rename both upscale and dpi variables accordingly.

bmachek commented:
> One other thing is that we like to include testing of new code, so if you could add that, eg https://github.com/paperless-ngx/paperless-ngx/blob/dev/src/documents/tests/test_barcodes.py

Added a new PDF in samples and wrote a pytest.

shamoon commented:
Thanks, I would perhaps include the before / after in that test (i.e. that it fails without the upscaling)

bmachek committed:
Added another pytest for barcode detection with dpi and upscale settings.
test_scan_file_for_qrcode_without_upscale() runs with 300 dpi and without upscaling.
It should fail detecting the ASN barcode in barcode-qr-asn-000123-upscale-dpi.pdf

bmachek commented:
> Thanks, I would perhaps include the before / after in that test (i.e. that it fails without the upscaling)

Did that by writing another test, with default settings for dpi and upscale. Maybe this test will fail someday, when pyzbar gets better or zxing becomes default. But IMHO it's OK for now.

shamoon commented on paperless.conf.example:
The new vars should have their default values here

shamoon commented on src/documents/tests/test_barcodes.py:
```diff
@@ -906,6 +906,45 @@ def test_asn_too_large(self):
 
+    def test_scan_file_for_qrcode_without_upscale(self):
+        """
+        ?????
```
```suggestion
```

shamoon commented on src/documents/tests/test_barcodes.py:
```diff
@@ -906,6 +906,45 @@ def test_asn_too_large(self):
+    @override_settings(CONSUMER_BARCODE_UPSCALE=1.5)
+    def test_scan_file_for_qrcode_with_upscale(self):
+        """
+        ?????
```
```suggestion
```

bmachek committed:
Added default values to PAPERLESS_CONSUMER_BARCODE_UPSCALE and PAPERLESS_CONSUMER_BARCODE_DPI in paperless.conf.example

Removed needless comments from test_barcodes.py
Forced the test methods test_scan_file_for_qrcode_with_upscale and test_scan_file_for_qrcode_without_upscale,
as the results are different with zxing used in CI. zxing doesn't seem to need any upscaling and dpi change
for the detection in test document: barcode-qr-asn-000123-upscale-dpi.pdf

bmachek commented:
I made the requested changes.

The CI runs using ZXing for my pytest methods show that the results differ for negative test without upscaling.
ZXing doesn't seem to need upscaling or higher dpi to detect the small QR code in the test PDF.
This puts the functionality upscaling/dpi to the state of a workaround for the build issues of ZXing on aarch64.
So it maybe removed once ZXing settles its problems there.

bmachek committed:
Merge branch 'paperless-ngx:dev' into dev

shamoon commented:
Yea thats part of why I wanted the testing, theres a bit of overlapping 'workarounds' here (making both barcode libraries available itself is a sort of workaround).

Thanks, perhaps you can throw in a comment to the tests to explain the above a bit (for some future date). Otherwise, Im good with this, though I will leave final approval to the backend folks (the main person of which is currently AFK so there may be a small delay in finalizing this).

Thanks again

bmachek committed:
Explained why pyzbar is used in test_barcodes.py ibn the comments.
Upscaling and dpi raise in conversion is a workaround, when you run into problems with barcode detection, and cannot use zxing. (zxing wheel aarch64).

bmachek committed:
Merge branch 'dev' of github.com:bmachek/paperless-ngx into dev

stumpylog reviewed (approved):
I'm good with the changes.  I'll leave it up the @shamoon if this merges to `dev` or if that happens a little later (after other bugfix releases)

shamoon commented:
Cheers all, thanks for all the work @bmachek
